compiler/luci/pass/src/RemoveUnnecessaryTransposeNetPass.cpp

```diff
@@ -360,14 +385,33 @@ bool TaggedShapeAnalyzer::can_remove_transposes()
   assert(_in != nullptr && _front_transpose != nullptr && _mid_reshape != nullptr &&
          _back_transpose != nullptr);
 
-  init_shape_with_tag(_in_shape_v);
-
-  analyze_transpose(_front_perm_v);
+  auto count_not_1_dim = [](std::vector<int32_t> &shape) -> int {
+    int count = 0;
+    for (auto i : shape)
+    {
+      if (i != 1)
+        count++;
+    }
+    return count;
+  };
 
-  if (not analyze_reshape(_mid_shape_v))
-    return false;
+  auto in_rank = count_not_1_dim(_in_shape_v);
+  auto out_rank = count_not_1_dim(_out_shape_v);
 
-  analyze_transpose(_back_perm_v);
+  if (in_rank >= out_rank)
+  {
+    init_shape_with_tag(_in_shape_v);
+    analyze_transpose(_front_perm_v);
+    CHECK_OR_FALSE(analyze_reshape(_mid_shape_v));
+    analyze_transpose(_back_perm_v);
+  }
+  else
+  {
+    init_shape_with_tag(_out_shape_v);
+    analyze_transpose(inverse_perm(_back_perm_v));
+    CHECK_OR_FALSE(analyze_reshape(_front_shape_v));
+    analyze_transpose(inverse_perm(_front_perm_v));
```

Comment:
(note) this is main logic. 

For rank expansion cases, it analyzes the nodes(pattern) in reverse order.
	* init tag with output shape
	* analyze back_transpose
	* analyze resshape in middle
	* analyze front transpose
	* verify the tags are still in order

Reply:
> For rank expansion cases, it analyzes the nodes(pattern) in reverse order.

I've imagined something like multi-level tags, but this code looks simpler 👍